L67-cpp-callback-api.md

```diff
@@ -0,0 +1,125 @@
+C++ callback-based asynchronous API
+----
+* Author(s): vjpai, sheenaqotj, yang-g, zhouyihaiding
+* Approver: markdroth
+* Status: Proposed
+* Implemented in: https://github.com/grpc/grpc/projects/12
+* Last updated: March 22, 2021
+* Discussion at https://groups.google.com/g/grpc-io/c/rXLdWWiosWg
+
+## Abstract
+
+Provide an asynchronous gRPC API for C++ in which the completion of RPC actions in the library will result in callbacks to user code,
+
+## Background
+
+Since its initial release, gRPC has provided two C++ APIs:
+
+* Synchronous API
+   - All RPC actions (such as unary calls, streaming reads, streaming writes, etc.) block for completion
+* Completion-queue-based asynchronous API
+   - Application associates each RPC action that it initiates with a tag
+   - The library performs each RPC action
+   - The library posts the tag of a completed action onto a completion queue
+   - The application must poll the completion queue to determine which asynchronously-initiated actions have completed
+
+The goal of the synchronous version is to be easy to program. However, this comes at the cost of high thread-switching overhead and high thread storage for systems with many concurrent RPCs. On the other hand, the asynchronous API allows the application full control over its threading and thus can scale further. The biggest problem with the asynchronous API is that it is just difficult to use. Server RPCs must be explicitly requested, RPC polling must be explicitly controlled by the application, lifetime management is complicated, etc. These have proved sufficiently difficult that the full features of the asynchronous API are basically never used by applications. Even if one can use the async API correctly, it also presents challenges in deciding how many completion queues to use and how many threads to use for polling them, as one can either optimize for reducing thread hops, avoiding stranding, reducing CQ contention, or improving locality. These goals are often in conflict and require substantial tuning.
+
+### Related proposals
+
+* The C++ callback API is built on top of a new [callback completion queue in core](https://github.com/grpc/proposal/pull/181)
```

Comment:
Is it actually still built on that?  Or did your callback alternative approach wind up using a different type of CQ instead?

Reply:
Good observation. It's built on 2 possible implementations, so I should list both. Will get back to this on Monday.